Enable dynamo traced test_forloop_goes_right_direction

mlazos opened the pull request:
Removed a bunch of skips, I also updated test_forloop_goes_right_direction to *not* use the closure when dynamo is tracing. The reason for this is that testing the disabled optimizer doesn't actually test anything.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #123469
* #123619
* #123406
* #124309
* #123405
* #123404
* #123324
* #123323
* __->__ #123322
* #123498



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

mlazos committed:
Enable test_forloop_goes_right_direction

[ghstack-poisoned]

janeyx99 commented on test/test_optim.py:
```diff
@@ -112,7 +112,15 @@ def closure():
             initial_value = closure().item()
             for _ in range(20):
-                loss = optimizer.step(closure)
+                if torch._dynamo.is_compiling() and optim_cls != torch.optim.LBFGS:
```
Use the optim_info.step_requires_closure flag instead. Can you add a separate test to test correctness of closures under torch.compile in test_compiled_optimizers.py (that it falls back and computes the same thing)?

janeyx99 reviewed (commented):
Could you add more detail that the enablement is for dynamo testing

mlazos committed:
Update on "Enable dynamo traced test_forloop_goes_right_direction"

Removed a bunch of skips, I also updated test_forloop_goes_right_direction to *not* use the closure when dynamo is tracing. The reason for this is that testing the disabled optimizer doesn't actually test anything.




[ghstack-poisoned]

mlazos committed:
Update on "Enable dynamo traced test_forloop_goes_right_direction"

Removed a bunch of skips, I also updated test_forloop_goes_right_direction to *not* use the closure when dynamo is tracing. The reason for this is that testing the disabled optimizer doesn't actually test anything.




[ghstack-poisoned]

janeyx99 commented on test/inductor/test_compiled_optimizers.py:
```diff
@@ -343,66 +342,75 @@ def test_correctness(self, device, dtype, optim_info):
-                def fn():
-                    def closure():
-                        loss = model_compiled(input).sum()
+        closure_options = (True,) if optim_info.step_requires_closure else (False, True)
```
Better to use parametrization over the test similar to with_lrsched here: https://github.com/pytorch/pytorch/blob/main/test/test_optim.py#L82

It is easier to read and will make it separate tests while allowing you to reuse the code

mlazos replied:
I would like to skip the parameterization for the requires closure optimizers though. Don't see a way to do that with parameterizations..

janeyx99 commented on test/inductor/test_compiled_optimizers.py:
```diff
@@ -343,66 +342,75 @@ def test_correctness(self, device, dtype, optim_info):
-                        return loss
-
-                    opt_compiled.step(closure)
+                        if optim_cls is SparseAdam:
```
Use the supports_only_sparse_grads flag

mlazos committed:
Update on "Enable dynamo traced test_forloop_goes_right_direction"

Removed a bunch of skips, I also updated test_forloop_goes_right_direction to *not* use the closure when dynamo is tracing. The reason for this is that testing the disabled optimizer doesn't actually test anything.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

mlazos committed:
Update on "Enable dynamo traced test_forloop_goes_right_direction"

Removed a bunch of skips, I also updated test_forloop_goes_right_direction to *not* use the closure when dynamo is tracing. The reason for this is that testing the disabled optimizer doesn't actually test anything.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

mlazos committed:
Update on "Enable dynamo traced test_forloop_goes_right_direction"

Removed a bunch of skips, I also updated test_forloop_goes_right_direction to *not* use the closure when dynamo is tracing. The reason for this is that testing the disabled optimizer doesn't actually test anything.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

mlazos committed:
Update on "Enable dynamo traced test_forloop_goes_right_direction"

Removed a bunch of skips, I also updated test_forloop_goes_right_direction to *not* use the closure when dynamo is tracing. The reason for this is that testing the disabled optimizer doesn't actually test anything.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

mlazos committed:
Update on "Enable dynamo traced test_forloop_goes_right_direction"

Removed a bunch of skips, I also updated test_forloop_goes_right_direction to *not* use the closure when dynamo is tracing. The reason for this is that testing the disabled optimizer doesn't actually test anything.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

mlazos committed:
Update on "Enable dynamo traced test_forloop_goes_right_direction"

Removed a bunch of skips, I also updated test_forloop_goes_right_direction to *not* use the closure when dynamo is tracing. The reason for this is that testing the disabled optimizer doesn't actually test anything.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

mlazos committed:
Update on "Enable dynamo traced test_forloop_goes_right_direction"

Removed a bunch of skips, I also updated test_forloop_goes_right_direction to *not* use the closure when dynamo is tracing. The reason for this is that testing the disabled optimizer doesn't actually test anything.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

mlazos committed:
Update on "Enable dynamo traced test_forloop_goes_right_direction"

Removed a bunch of skips, I also updated test_forloop_goes_right_direction to *not* use the closure when dynamo is tracing. The reason for this is that testing the disabled optimizer doesn't actually test anything.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

mlazos committed:
Update on "Enable dynamo traced test_forloop_goes_right_direction"

Removed a bunch of skips, I also updated test_forloop_goes_right_direction to *not* use the closure when dynamo is tracing. The reason for this is that testing the disabled optimizer doesn't actually test anything.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

mlazos committed:
Update on "Enable dynamo traced test_forloop_goes_right_direction"

Removed a bunch of skips, I also updated test_forloop_goes_right_direction to *not* use the closure when dynamo is tracing. The reason for this is that testing the disabled optimizer doesn't actually test anything.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

janeyx99 commented on test/test_optim.py:
```diff
@@ -112,7 +112,15 @@ def closure():
             initial_value = closure().item()
             for _ in range(20):
-                loss = optimizer.step(closure)
+                if torch._dynamo.is_compiling() and not optim_info.step_requires_closure:
+                    # we disable torch.compile with the closure argument, so in order to actually
+                    # test compiling the optimizer, we separate out the closure here.
+                    # LBFGS requires the closure arg but torch.compile doesn't support LBFGS
+                    # so it's fine to run the disabled path.
+                    loss = closure()
+                    optimizer.step()
+                else:
+                    loss = optimizer.step(closure)
```
It's fine to just flip this to
```suggestion
                if optim_info.step_requires_closure:
                    loss = optimizer.step(closure)
                else:
                    # we disable torch.compile with the closure argument, so in order to actually
                    # test compiling the optimizer, we separate out the closure here.
                    loss = closure()
                    optimizer.step()
                    
```

janeyx99 reviewed (approved):
Approving, but use my suggestion to minimize disparity between eager and dynamo tests,

mlazos committed:
Update on "Enable dynamo traced test_forloop_goes_right_direction"

Removed a bunch of skips, I also updated test_forloop_goes_right_direction to *not* use the closure when dynamo is tracing. The reason for this is that testing the disabled optimizer doesn't actually test anything.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

mlazos committed:
Update on "Enable dynamo traced test_forloop_goes_right_direction"

Removed a bunch of skips, I also updated test_forloop_goes_right_direction to *not* use the closure when dynamo is tracing. The reason for this is that testing the disabled optimizer doesn't actually test anything.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

mlazos committed:
Update on "Enable dynamo traced test_forloop_goes_right_direction"

Removed a bunch of skips, I also updated test_forloop_goes_right_direction to *not* use the closure when dynamo is tracing. The reason for this is that testing the disabled optimizer doesn't actually test anything.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

mlazos committed:
Update on "Enable dynamo traced test_forloop_goes_right_direction"

Removed a bunch of skips, I also updated test_forloop_goes_right_direction to *not* use the closure when dynamo is tracing. The reason for this is that testing the disabled optimizer doesn't actually test anything.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

mlazos commented:
@pytorchbot merge

mlazos committed:
Update on "Enable dynamo traced test_forloop_goes_right_direction"

Removed a bunch of skips, I also updated test_forloop_goes_right_direction to *not* use the closure when dynamo is tracing. The reason for this is that testing the disabled optimizer doesn't actually test anything.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

mlazos commented:
@pytorchbot merge
